synapse/util/caches/__init__.py

```diff
@@ -195,8 +195,10 @@ def register_cache(
     )
 }
 
+T = TypeVar("T", Optional[str], str)
```

Comment:
This is attempting to make it so that if we know we have a `str` before calling `intern_string` we don't lose that information (and have it become `Optional[str]` on the output). I'm not 100% sure this is correct though.

So pretty much `intern_string(Optional[str]) -> Optional[str]` while `intern_string(str) -> str`. Could maybe also be done with an overload?

Reply:
> Could maybe also be done with an overload?

For sure. I don't mind this way though: [typing.AnyStr](https://docs.python.org/3/library/typing.html?highlight=anystr#typing.AnyStr) is a nice example of this in the stdlib.

With that said: is there an example where this function is passed `None`? I guess `intern_dict` could have a key which maps to `None`?

Reply:
(I was gonna say "a name more specific than `T`" might be nice. But I couldn't think of a good one. `FixedOptionalString`?

Reply:
> With that said: is there an example where this function is passed `None`? I guess `intern_dict` could have a key which maps to `None`?

I think the only place it could be is the `intern_dict` call for the state key? I'm not 100% sure though...

Reply:
Nor I! Let's leave it be for now.